src/dtx/dtx_common.c

```diff
@@ -1204,33 +1212,39 @@
 	if (daos_is_zero_dti(&dth->dth_xid))
 		D_GOTO(out, result = result < 0 ? result : rc);
 
-	if (dth->dth_need_validation) {
+	if (result < 0 || rc < 0)
+		D_GOTO(abort, result = result < 0 ? result : rc);
+
+	if (dth->dth_need_validation || unlikely(dth->dth_aborted)) {
 		/* During waiting for bulk data transfer or other non-leaders, the DTX
 		 * status may be changes by others (such as DTX resync or DTX refresh)
 		 * by race. Let's check it.
 		 */
 		status = vos_dtx_validation(dth);
-		if (unlikely(status == DTX_ST_COMMITTED || status == DTX_ST_COMMITTABLE))
+		switch (status) {
+		case DTX_ST_PREPARED:
+			if (likely(!dth->dth_aborted))
+				break;
+			/* Fall through */
+		case DTX_ST_INITED:
+		case DTX_ST_PREPARING:
+			aborted = true;
+			D_GOTO(out, result = -DER_AGAIN);
+		case DTX_ST_ABORTED:
+		case DTX_ST_ABORTING:
+			aborted = true;
+			D_GOTO(out, result = -DER_INPROGRESS);
+		case DTX_ST_COMMITTED:
+		case DTX_ST_COMMITTING:
```

Comment:
(style)  Possible switch case/default not preceeded by break or fallthrough comment